src/os/bluestore/BlueStore.cc

```diff
@@ -9501,9 +9501,18 @@ void BlueStore::_kv_sync_thread()
 	}
       }
 
-      // submit synct synchronously (block and wait for it to commit)
-      int r = cct->_conf->bluestore_debug_omit_kv_commit ? 0 : db->submit_transaction_sync(synct);
-      ceph_assert(r == 0);
+      RocksDBStore::RocksDBTransactionImpl * _t =
+        static_cast<RocksDBStore::RocksDBTransactionImpl *>(synct.get());
+      RocksDBStore::RocksWBHandler bat_txc;
+      _t->bat.Iterate(&bat_txc);
+
```

Comment:
Iterating over submit batch just to check if it is empty is problematic, especially because kv_sync_thread is known bottleneck. Moreover, synct can be empty, but we could have already submitted some other batches to db ( _txc_apply_kv).
We should gather information whether we need to submit sync by other means.

Reply:
@aclamk  Thanks for your review.

In addition to the judgment of synct  is empty, there is also the judgment of whether the bluefs_do_check_balance is true.

As you konw，when the bluefs_do_check_balance is true，the kv_queue should be empty , so we impossibility have already submitted some other batches to db. Right?

Thanks again for your review.

Reply:
@wuhongsong 
1) "As you konw，when the bluefs_do_check_balance is true，the kv_queue should be empty , so we impossibility have already submitted some other batches to db."
I did not actually realized that. True, each time we add to kv_queue we signal kv_cond, so it IS impossible to get both timeout and non-empty kv_queue.
I am only afraid that relying on this might bring us an error when we "optimize" notification frequency, as we did with `deferred_done_queue`.
2) Iterating over bat_txc should be done only when bluefs_do_check_balance is true. Unless there any other situation for us to get empty transaction; which I do not see.